JamesAGarcia/liferay-portal
# LRDOCS-11747 adds liferay-course-workspace w/client extensions

## Contents

I added the liferay-course-workspace, which includes two projects:  `liferay-course-batch` and `liferay-course-etc-spring-boot`.

* The batch project includes two JSON files.
   *  The first is for a simplified Distributor Application object (used for testing only)
   *  The second is for a workflow process that's used by the custom workflow action.
* The spring boot project includes a custom workflow action and object action. 
   * The workflow action determines the transition based on a field value in the object entry
   * The object action chains three async http post requests:
      * The first post request creates a business account
      * The second associate a user with the account
      * The third assigns the user an account role

## Testing the Object Action

1. Navigate to the `/liferay-portal/workspaces/liferay-course-workspace` folder and execute these commands:

   ```bash
   ./gradlew :client-extensions:liferay-course-etc-spring-boot:deploy
   ```

   ```bash
   ./gradlew :client-extensions:liferay-course-batch:deploy
   ```

1. Move the generated zips from `/liferay-portal/workspaces/liferay-course-workspace/bundles/osgi/client-extensions` to the `/bundles/osgi/client-extensions` folder. 

1. After they finish starting, execute this command from `liferay-sample-workspace`:

   ```bash
   ./gradlew :client-extensions:liferay-sample-etc-spring-boot:bootRun
   ```

1. Verify the `Distributor Application` object definition was created. It should include an action for calling the custom action On After Add.

1. Create an entry using test@liferay.com for applicant email and whatever you want for business name.

    (NOTE: If the Distributor Application is inaccessible via the global menu, deactivate and reactivate the object definition. That should fix it. This started happening after I added the action to the definition's JSON.)

1. Open the Accounts application and verify there is a new account that uses the business  name.

1. Open the account, go to the Users tab, and verify the test@liferay.com user is associated with the account.

1. Go to the Roles tab, select the Account Administrator role, and verify the test@liferay.com user is assigned the role.

## Testing the Workflow Action

To test the workflow action, you need to prepare the Distributor Application object:

1. Create a picklist called _States_.

1. Add these values to the States picklist: _Open_, _Approved_, and _Denied_.

1. Use the picklist as an object field called _State_.

1. If you haven't already, deploy the client extensions and start up the springboot server (steps 1-3 of Testing the Object Action).

1. Open the Global Menu, go to the Applications tab, and click Process Builder.

1. Confirm the Machine Approver process appears.

1. Go to the Configuration tab.

1. Click Edit for Distributor Application, select Machine Approver, and click Save.

1. Open the Distributor Application object, create an entry with the 'open' state and check its status. It should be approved.

1. Update the entry's state to Approved and check its status. It should be In Review.

1. Update the entry's state back to Open and check its status. It should be approved.
